### PR TITLE
fix(DTFS-6107): partyURN validation

### PR DESCRIPTION
**Pull request by Zainzzkk**
# Changes made:

* Added partyURN validation so that it is more than 3 numbers long and only numbers
* Added error summary and inline error for partyURN input
* Added unit and e2e tests

### CODE REVIEW

File: e2e-tests/trade-finance-manager/cypress/e2e/journeys/case/parties/exporter.spec.js [hunk 1]
```diff
@@ -61,6 +61,48 @@ context('Parties - user can view and edit exporter', () => {
 
         pages.exporterPage.saveButton().click();
 
+        cy.url().should('eq', relative(`/case/${dealId}/parties/exporter`));
+        pages.exporterPage.errorSummary().contains('Enter a correct party URN');
+        pages.exporterPage.urnError().contains('Enter a correct party URN');
```
Comment (abhi-markan) [inline e2e-tests/trade-finance-manager/cypress/e2e/journeys/case/parties/exporter.spec.js:66]: Can we please add a test case with empty input please.

File: trade-finance-manager-ui/server/controllers/case/parties/index.test.js [hunk 2]
```diff
@@ -125,6 +125,7 @@ describe('controllers - case - parties', () => {
           tfm: mockDeal.tfm,
           dealId: req.params._id,
           user: session.user,
+          urn: mockDeal.tfm.parties.exporter.partyUrn,
```
Comment (abhi-markan) [inline trade-finance-manager-ui/server/controllers/case/parties/index.test.js:128]: This test is only concentrated at `exporter` party, other parties test should also be incorporated:

* Agent
* Indemnifier
* Bond Issuer
* Bond Beneficiary

File: trade-finance-manager-ui/server/controllers/case/parties/index.test.js [hunk 3]
```diff
@@ -353,16 +354,113 @@ describe('controllers - case - parties', () => {
             _id: mockDeal._id,
           },
           body: {
-            exporter: {
-              partyUrn: '12345',
-            },
+            partyUrn: '12345',
```
Comment (abhi-markan) [inline trade-finance-manager-ui/server/controllers/case/parties/index.test.js:357]: Should this be not inside `exporter` like other parties?

File: trade-finance-manager-ui/server/controllers/case/parties/partyUrnValidation.validate.test.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  it('should return validation error if partyUrn is letters and numbers', () => {
+    const urn = '12C';
+
+    const response = validatePartyURN(urn);
+    expect(response.errorsObject.errors.errorSummary).toEqual([{ text: 'Enter a correct party URN', href: '#partyUrn' }]);
+    expect(response.errorsObject.errors.fieldErrors).toEqual({ partyUrn: { text: 'Enter a correct party URN' } });
+    expect(response.urn).toEqual('12C');
+  });
+
+  it('should return validation error if partyUrn is correct length of numbers but has a special character', () => {
```
Comment (abhi-markan) [inline trade-finance-manager-ui/server/controllers/case/parties/partyUrnValidation.validate.test.js:83]: Can we please add a case with an empty space input please.

File: trade-finance-manager-ui/server/controllers/case/parties/index.js [hunk 5]
```diff
@@ -145,6 +147,34 @@ const postPartyDetails = (partyType) => (
       return res.redirect('/not-found');
     }
 
+    /**
```
Comment (abhi-markan) [inline trade-finance-manager-ui/server/controllers/case/parties/index.js:150]: Can we please take into account other parties.

File: trade-finance-manager-ui/server/controllers/case/parties/index.js [hunk 6]
```diff
@@ -145,6 +147,34 @@ const postPartyDetails = (partyType) => (
...
+     * checks if partyType is exporter and if partyUrn input exists or is a blank string (nothing inputted)
+     * validates input - should only be at least 3 numbers
+     * validation error rendered if error exists on edit partyUrn page
+     * if no error, then updates partyURN
+     */
+    if (partyType === 'exporter' && (req.body.partyUrn || req.body.partyUrn === '')) {
```
Comment (abhi-markan) [inline trade-finance-manager-ui/server/controllers/case/parties/index.js:156]: You can use a helper function for string which ascertain whether string exists and is not null or undefined

**Review by abhi-markan — CHANGES_REQUESTED** [review]

File: trade-finance-manager-ui/server/controllers/case/parties/index.test.js (hunk 2 above)
Comment (Zainzzkk) [reply trade-finance-manager-ui/server/controllers/case/parties/index.test.js:128]: unable to add these but e2e tests cover these

File: trade-finance-manager-ui/server/controllers/case/parties/index.test.js (hunk 3 above)
Comment (Zainzzkk) [reply trade-finance-manager-ui/server/controllers/case/parties/index.test.js:364]: discussed.  Sits outside in the body just as partyUrn

File: trade-finance-manager-ui/server/controllers/case/parties/index.js (hunk 6 above)
Comment (Zainzzkk) [reply trade-finance-manager-ui/server/controllers/case/parties/index.js]: done

File: trade-finance-manager-ui/server/controllers/case/parties/index.js (hunk 5 above)
Comment (Zainzzkk) [reply trade-finance-manager-ui/server/controllers/case/parties/index.js:150]: done

File: trade-finance-manager-ui/server/controllers/case/parties/partyUrnValidation.validate.test.js (hunk 4 above)
Comment (Zainzzkk) [reply trade-finance-manager-ui/server/controllers/case/parties/partyUrnValidation.validate.test.js:107]: done

File: e2e-tests/trade-finance-manager/cypress/e2e/journeys/case/parties/exporter.spec.js (hunk 1 above)
Comment (Zainzzkk) [reply e2e-tests/trade-finance-manager/cypress/e2e/journeys/case/parties/exporter.spec.js]: done

File: trade-finance-manager-ui/server/controllers/case/parties/helpers.js [hunk 7]
```diff
@@ -3,4 +3,50 @@ const { userIsInTeam } = require('../../../helpers/user');
...
+
+  if (isBondPartyType(partyType)) {
+    errRef = `partyUrn-${index}`;
+  }
+
+  return errRef;
+};
+
+const isEmptyString = (str) => {
+  if (!str || ((typeof str === 'string' || str instanceof String) && !str.trim().length)) {
```
Comment (abhi-markan) [inline trade-finance-manager-ui/server/controllers/case/parties/helpers.js:40]: Do we need to make `typeof` and `instaceof` condition an `&&` condition. As you can pass an `undefined` object which will return true.

Is the function only expecting String typeof?

File: trade-finance-manager-ui/server/controllers/case/parties/index.test.js (hunk 2 above)
Comment (abhi-markan) [reply trade-finance-manager-ui/server/controllers/case/parties/index.test.js:128]: Sorry, can you please remind me again why are we unable to add these here?

**Review by abhi-markan — CHANGES_REQUESTED** [review]
Some good work there, few comments to address please.

File: trade-finance-manager-ui/server/controllers/case/parties/index.test.js (hunk 2 above)
Comment (Zainzzkk) [reply trade-finance-manager-ui/server/controllers/case/parties/index.test.js:128]: done

File: trade-finance-manager-ui/server/controllers/case/parties/helpers.js (hunk 7 above)
Comment (Zainzzkk) [reply trade-finance-manager-ui/server/controllers/case/parties/helpers.js:41]: discussed, leaving as is as will never be undefined or null and validation can handle

File: trade-finance-manager-ui/server/controllers/case/parties/helpers.js [hunk 8]
```diff
@@ -3,4 +3,50 @@ const { userIsInTeam } = require('../../../helpers/user');
...
+  let errRef = 'partyUrn';
+
+  if (isBondPartyType(partyType)) {
+    errRef = `partyUrn-${index}`;
+  }
+
+  return errRef;
+};
+
+const isEmptyString = (str) => {
```
[the remark this answers is not in the document]
Comment (Zainzzkk) [reply trade-finance-manager-ui/server/controllers/case/parties/helpers.js:40]: done

File: trade-finance-manager-ui/server/controllers/case/index.js [hunk 9]
```diff
@@ -352,13 +354,59 @@ const postTfmFacility = async (req, res) => {
   const dealId = req.params._id;
```
[the remark this answers is not in the document]
Comment (Zainzzkk) [reply trade-finance-manager-ui/server/controllers/case/index.js:359]: done

File: trade-finance-manager-ui/server/controllers/case/parties/helpers.js (hunk 7 above)
Comment (Zainzzkk) [reply trade-finance-manager-ui/server/controllers/case/parties/helpers.js:41]: discussed, leaving as is as will never be undefined or null and validation can handle

**Review by abhi-markan — APPROVED** [review]
